graftm/taxonomy.py

```diff
@@ -0,0 +1,176 @@
+        
+        Parameters
+        ----------
+        taxa_id : array
+            taxa id to return the taxonomy for
+        '''
+        if taxa_id in self.id_set:
+            return True
+        else:
+            return False
+        
+    def entry_ids(self):
+        '''
+        Return a list of all taxonomy entries within the self.taxnomy dict
+        '''
+        return self.taxonomy.keys()
+    
+    def rank(self, taxa_id, rank):
+        '''
+        Parameters
+        ----------
+        taxa_id : array
+            taxa id to return the taxonomy for
+        rank : int
+            rank to return
+        Returns
+        -------
+        rank as indicated by the index passed to 'rank'.
+        '''
+        if rank < self.max:
+            return self.taxonomy[taxa_id][rank]
+        else:
+            raise ProgrammingError("The rank specified was larger than the \
+maximum possible number of ranks within the taxonomy description.")
+        
+        
+    
+    def taxonomy_string(self, taxa_id):
+        '''
+        Return the tax string of the id provided
+        
+        Parameters
+        ----------
+        taxa_id : array
+            taxa id to return the taxonomy for
+        
+        Returns
+        -------
+        The taxonomy string joined with the standard self.GG_SEP_0 separator.
+        '''
+        
+        if taxa_id in self.taxonomy:
+            taxonomy_string = GG_SEP_0.join(self.taxonomy[taxa_id])
+        else:
+            raise MissingTaxonomy("Taxa id provided did not have any \
+any corresponding entry in the taxonomy description.")
+            
+        return taxonomy_string
+            
+    def read_format_greengenes(self, taxonomy_io):
+        '''
+        Parses a greengenes taxonomy file. No set number of taxonomic ranks is 
+        required
+        
+        Parameters
+        ----------
+        taxonomy_io : Open greengenes taxonomy file
+        '''
+        logging.debug("Reading in greengenes formatted taxonomy")
+        try:
+            for entry in taxonomy_io:
+                entry_id, entry_taxonomy_string = entry.strip()\
+                                                       .split(self.TAB)
+                
+                if self.GG_SEP_0 in entry_taxonomy_string:
+                    split_entry_taxonomy = \
+                                entry_taxonomy_string.split(self.GG_SEP_0)                
+                else:
+                    split_entry_taxonomy = \
+                                 entry_taxonomy_string.split(self.GG_SEP_1)
+                
+                self.taxonomy[entry_id] = split_entry_taxonomy
+            self._ammend_taxonomy_gaps()
+            self.id_set = set(self.taxonomy.keys())
+        except:
+            logging.error("Controlled exit from taxonomy.Taxonomy \
+read_format_greengenes")
+            raise MalformedGreenGenesFormattedFile("The GreenGenes taxonomy file \
```

Comment:
Be good to include the line that was malformed, helps the user fix their GG file.